### PR TITLE
New parser

### PR DESCRIPTION
**Pull request by hit9**
**Dont Merge Now.**

New parser with bugs fixes:

1. #69 constant value from referenced constant or enum.
2.  #75 similar to #69, but requires typedefs && ttypes (as field_type)
3.  Enable optional semicolon at the end of a head_unit or definition_unit.
4.  Remove union (wont support)

e.g.

1. 
   ```thrift
   enum Country {
       CN = 1
   }

   struct Person {
       1: Country country = Country.CN
   }
   ```
   => 

   ```python
   {
     "ttypes": {
         "Country": {
         "CN": 1
      },
       "Person": {
          "country": {
           "id": 1,
           "required": true,
           "type": "Country",
           "value": 1
         }
       }
     }
   }
   ```

Todo steps:

* tests
* typedefs as a part of  field_type

In-plan unsupported parsing:

```thrift
struct Persion {
  1: string name
  2: i32 age
}

const Person liming = {"name": "liming", "age": 17}
```

```thrift
typedef i32 Int
const Int myint = 32
```

### CODE REVIEW

**Comment by hit9** [comment]
和 @maralla 讨论后，parser/__init__ 中的load部分直接分散到ply的p_函数中去，即parser.parse直接返回一个thrift module对象。

**Comment by hit9** [comment]
1. Global `thrift` => Globa `thrifts_stack` (for includes)
2. Deprecate `namespace`.

**Comment by hit9** [comment]
3. Rename include_dirs => include_dir (like gcc way: gcc filepath -Iinclude-path)

File: thriftpy/parser/parser.py [hunk 1]
```diff
@@ -10,17 +10,7 @@
...
-    Field,
-    Function,
-    IdentifierValue,
-    ListType,
-    MapType,
-    Service,
-    SetType,
-    Thrift
-)
+from .types import *  # noqa
```
Comment (tmehlinger) [inline thriftpy/parser/parser.py]: Please don't do this. It's much easier to reason about the code with explicit imports.

**Comment by tmehlinger** [comment]
Why are you removing union support? This is a significant breaking change that could negatively impact current thriftpy users.

File: thriftpy/parser/parser.py (hunk 1 above)
Comment (hit9) [inline thriftpy/parser/parser.py]: OK, this dose not matter but I will use `explicit imports` if you like.

**Comment by hit9** [comment]
@tmehlinger We mixed structs/unions for now, but they are total different things. 

But apache thrift just dose the same thing: process unions the same with structs.

Thanks and I think we shouldn't drop unions and should keep them the same with structs (In python)

**Comment by hit9** [comment]
Reopen on https://github.com/eleme/thriftpy/pull/80

**Pull request closed without merging**